File: Python/Archive/871. Minimum Number of Refueling Stops.py
import heapq

class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        heap = []
        stations.append([target, 0])
        ans = 0
        prev = 0
        currFuel = startFuel
        for position, fuel in stations:
            currFuel -= (position - prev)
            while currFuel < 0 and len(heap) > 0:
                currFuel += -heapq.heappop(heap)
                ans += 1
            if currFuel < 0:
                return -1
            heapq.heappush(heap, -fuel)
            prev = position
        return ans
        

File: C++/Heap (Priority Queue)/871. Minimum Number of Refueling Stops.cpp
class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        priority_queue<int> pq;
        stations.push_back({target, 0});
        int res = 0, prev = 0;
        for (auto& station : stations) {
            int location = station[0], capacity = station[1];
            startFuel -= location - prev;
            while (startFuel < 0 && !pq.empty()) {
                startFuel += pq.top();
                pq.pop();
                res++;
            }
            if (startFuel < 0) return -1;
            pq.push(capacity);
            prev = location;
        }
        return res;
    }
};